src/classes/BGE_DataImportBatchEntry_CTRL.cls

```diff
@@ -125,6 +125,51 @@ public with sharing class BGE_DataImportBatchEntry_CTRL {
         }
     }
 
+    /*******************************************************************************************************
+    * @description Retrieves all open Opportunities and unpaid Payments for the selected Donor (Account or Contact)
+    * @return JSON of data model TBD
+    */
+    @AuraEnabled
+    public static String getOpenDonations(Id donorId, String donorType) {
+        String relationshipField;
+        if (donorType == 'Account1') {
+            relationshipField = 'AccountId';
+        } else if (donorType == 'Contact1') {
+            relationshipField = 'Primary_Contact__c';
+        }
+
+        String opportunitySOQL = 'SELECT Id, ' +
+                                'Name, ' +
+                                'StageName, ' +
+                                'Amount ' +
+                                'FROM Opportunity ' +
+                                'WHERE ' + relationshipField + ' = :donorId ' +
+                                'AND IsClosed = false';
+
+        String paymentSOQL = 'SELECT Id, ' +
+                                'Name, ' +
+                                'npe01__Scheduled_Date__c, ' +
+                                'npe01__Opportunity__r.Name, ' +
+                                'npe01__Opportunity__c, ' +
+                                'npe01__Payment_Amount__c ' +
+                                'FROM npe01__OppPayment__c ' +
+                                'WHERE npe01__Opportunity__r.' + relationshipField + ' = :donorId ' +
+                                'AND npe01__Paid__c = false';
+
+        try {
+            List<Opportunity> openOpportunities = Database.query(opportunitySOQL);
+            List<npe01__OppPayment__c> unpaidPayments = Database.query(paymentSOQL);
```

Comment:
Could we run the payment query first, build a set of Ids of related opps from that query, and use it to filter out the opp query?

```
List<npe01__OppPayment__c> unpaidPayments = Database.query(paymentSOQL);

Set<Id> oppsWithUnpaidPayments = new Set<Id>();
for (npe01__OppPayment__c unpaidPayment : unpaidPayments) {
    oppsWithUnpaidPayments.add(unpaidPayment. npe01__Opportunity__c;
}

if (oppsWithUnpaidPayments != null) { 
    opportunitySOQL += ' AND Id NOT IN :oppsWithUnpaidPayments';
}
```

Or we could consider removing the opps from the list after query if performance is a problem.

Reply:
Let's go with querying for all opportunities and removing those with payments after the query. The easiest way I can think to do this is to create a map of opportunities from the query results so you can easily remove keys by Id.

Reply:
^ @mdeleon0

Reply:
@bethbrains @randi274 Double check my thought process here, please :)